xapian-letor/tests/api_letor.cc

```diff
@@ -804,3 +804,42 @@ DEFINE_TESTCASE(err_scorer, !backend)
 
     return true;
 }
+
+//Test for TfDoclenFeature
+DEFINE_TESTCASE(tfdoclenfeature, generated)
+{
+    string db_path = get_database_path("apitest_tfdoclenfeature",
+				       db_index_one_document);
+    Xapian::Database db(db_path);
+    Xapian::Document doc = db.get_document(1);
+    std::map<std::string,Xapian::termcount> len;
+    Xapian::TermIterator dt = doc.termlist_begin();
+    
+    len["title"] = 4;
+    Xapian::termcount whole_len = db.get_doclength(1);
+    len["whole"] = whole_len;
+    TEST_EQUAL(whole_len,186);
+    len["body"] = whole_len - len["title"];
+
+    Xapian::TfDoclenFeature feature;
+    feature.set_doc_length(len);
+
+    std::map<std::string, Xapian::termcount> tf;
+    tf["ZStiger"] = 1;
+    tf["ZXDtiger"] = 6;
+    tf["Ztiger"] = 2;
+    feature.set_termfreq(tf);
+
+    vector<double> res = feature.get_values();
+    TEST_EQUAL(res.size(), 3);
+
+    //test for title
+    TEST_EQUAL(res[0], log10(1 + (1 / (1 + len["title"]))));
+
+    //test for body
+    TEST_EQUAL(res[1], log10(1 + (18 / (1 + len["body"]))));
+
+    //test for whole
+    TEST_EQUAL(res[2], log10(1 + (19 / (1 + len["whole"]))));
```

Comment:
The calculation of the expected answers here will use integer division, so e.g. the first is `log10(1 + (1 / (1 + len["title"]))` which is `log10(1 + (1 / (1 + 4))` but `1 / (1 + 4)` is zero in integer division so that's `log10(1 + 0)` which is just `0`.

The simplest way to fix this is to use `1.0 + len["title"]` since `1.0` is a `double` constant, and then everything gets calculated as floating point.

So if this test passes currently then there's something wrong somewhere!

I'd also tend to suggest stylistically that `log10(1 + 1 / (1 + len["title"])))` is a little clearer than `log10(1 + (1 / (1 + len["title"]))))` - you can reasonably assume that someone reading the code understands the relative precedence of addition/subtraction vs multiplication/division (it's the same in almost all programming languages, and matches the precedence in mathematics), but adding extra parentheses makes it harder for someone reading the code to mentally pair them up.

Reply:
> So if this test passes currently then there's something wrong somewhere!
>
sorry, I forgot to set the query for the feature. I will fix that.
> The calculation of the expected answers here will use integer division
> 
I missed that, will fix it.

Reply:
> sorry, I forgot to set the query for the feature. I will fix that.

Aha, that would explain it.